throw more specific exceptions in the python SDK

In a few places we're throwing `Exception`:

```
$ make lint
pylint --output-format=colorized examples/ readme_metrics/ --ignore=venv
************* Module readme_metrics.PayloadBuilder
readme_metrics/PayloadBuilder.py:266:12: W0719: Raising too general exception: Exception (broad-exception-raised)
readme_metrics/PayloadBuilder.py:315:8: W0719: Raising too general exception: Exception (broad-exception-raised)
************* Module readme_metrics.VerifyWebhook
readme_metrics/VerifyWebhook.py:9:12: W0719: Raising too general exception: Exception (broad-exception-raised)
readme_metrics/VerifyWebhook.py:21:12: W0719: Raising too general exception: Exception (broad-exception-raised)
readme_metrics/VerifyWebhook.py:36:12: W0719: Raising too general exception: Exception (broad-exception-raised)
************* Module readme_metrics.publisher
readme_metrics/publisher.py:98:12: W0719: Raising too general exception: Exception (broad-exception-raised)
```

where we really ought to be throwing something more specific.

I'm going to add this to the list of things not to check for now, but I'm filling this issue so I can come back and fix it.

see #848 for context

--- a/packages/python/.pylintrc
+++ b/packages/python/.pylintrc
@@ -13,4 +13,3 @@ disable=
     R0913, # too-many-arguments
     W0201, # attribute-defined-outside-init
     W0703, # broad-except
-    W0719, # broad-exception-raised (see https://github.com/readmeio/metrics-sdks/issues/849)

--- a/packages/python/readme_metrics/PayloadBuilder.py
+++ b/packages/python/readme_metrics/PayloadBuilder.py
@@ -13,6 +13,14 @@ import uuid
 from readme_metrics import ResponseInfoWrapper
 
 
+class QueryNotFound(Exception):
+    pass
+
+
+class BaseURLError(Exception):
+    pass
+
+
 class PayloadBuilder:
     """
     Internal builder class that handles the construction of the request and response
@@ -263,7 +271,7 @@ class PayloadBuilder:
             # works for Django, and possibly other request objects too
             result = request.environ["QUERY_STRING"]
         else:
-            raise Exception(
+            raise QueryNotFound(
                 "Don't know how to retrieve query string from this type of request"
             )
 
@@ -312,7 +320,7 @@ class PayloadBuilder:
                 return f"{scheme}://{host}{path}?{query_string}"
             return f"{scheme}://{host}{path}"
 
-        raise Exception("Don't know how to build URL from this type of request")
+        raise BaseURLError("Don't know how to build URL from this type of request")
 
     # always returns a dict with some of these fields: text, mimeType, params
     def _process_body(self, content_type, body):

--- a/packages/python/readme_metrics/VerifyWebhook.py
+++ b/packages/python/readme_metrics/VerifyWebhook.py
@@ -3,10 +3,14 @@ import hmac
 from datetime import datetime, timedelta
 
 
+class VerificationError(Exception):
+    pass
+
+
 class VerifyWebhook:
     def __init__(self, body: dict, signature: str, secret: str):
         if signature is None:
-            raise Exception("Missing Signature")
+            raise VerificationError("Missing Signature")
 
         parsed_input = dict(
             (x.strip(), y.strip())
@@ -18,7 +22,7 @@ class VerifyWebhook:
         if datetime.now() - datetime.fromtimestamp(
             int(parsed_input["t"]) / 1000
         ) > timedelta(minutes=30):
-            raise Exception("Expired Signature")
+            raise VerificationError("Expired Signature")
 
         unsigned = time + "." + json.dumps(body, separators=(",", ":"))
         verify_signature = hmac.new(
@@ -33,4 +37,4 @@ class VerifyWebhook:
             )
             is False
         ):
-            raise Exception("Invalid Signature")
+            raise VerificationError("Invalid Signature")

--- a/packages/python/readme_metrics/publisher.py
+++ b/packages/python/readme_metrics/publisher.py
@@ -15,6 +15,10 @@ BACKOFF_SECONDS = 15
 backoff_expires_at = None
 
 
+class HTTPError(Exception):
+    pass
+
+
 def start_backoff():
     global backoff_expires_at  # pylint: disable=global-statement
     # Backoff for a few seconds, but not if another concurrent request has
@@ -95,7 +99,7 @@ def publish_batch(config, queue):
 
         if not readme_result.ok:
             config.LOGGER.exception(readme_result.text)
-            raise Exception(f"POST to {url} returned {readme_result.status_code}")
+            raise HTTPError(f"POST to {url} returned {readme_result.status_code}")
     except Exception as e:
         # Errors in the Metrics SDK should never cause the application to
         # throw an error. Log it but don't re-raise.

